constraint/pkg/client/client.go

```diff
@@ -236,59 +270,121 @@ func (c *Client) CreateCRD(ctx context.Context, templ *templates.ConstraintTempl
+		return nil, err
+	}
 
 	req := ruleArities{
 		"violation": 1,
 	}
-	if err := requireRules(path, src, req); err != nil {
+	if err := requireRulesModule(entryPoint, req); err != nil {
 		return nil, fmt.Errorf("Invalid rego: %s", err)
 	}
 
-	_, err = c.rConformer.ensureRegoConformance(crd.Spec.Names.Kind, path, src)
+	rr.AddEntryPointModule(entryPointPath, entryPoint)
+	for idx, libSrc := range targetSpec.Libs {
+		libPath := fmt.Sprintf(`%s["lib_%d"]`, libPrefix, idx)
+		if err := rr.AddLib(libPath, libSrc); err != nil {
+			return nil, err
+		}
+	}
+
+	sources, err := rr.Rewrite()
 	if err != nil {
 		return nil, err
 	}
 
-	return crd, nil
+	var mods []constraintTemplateModule
+	if err := sources.ForEachModule(func(m *regorewriter.Module) error {
+		content, err := m.Content()
+		if err != nil {
+			return err
+		}
+		mods = append(mods, constraintTemplateModule{
+			path: m.Module.Package.Path.String(),
+			rego: string(content),
+		})
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return &constraintTemplateArtifacts{
+		crd:           crd,
+		targetHandler: targetHandler,
+		modules:       mods,
+	}, nil
+}
+
+// CreateCRD creates a CRD from template
+func (c *Client) CreateCRD(ctx context.Context, templ *templates.ConstraintTemplate) (*apiextensions.CustomResourceDefinition, error) {
+	artifacts, err := c.createTemplateArtifacts(templ)
+	if err != nil {
+		return nil, err
+	}
+	return artifacts.crd, nil
 }
 
 // AddTemplate adds the template source code to OPA and registers the CRD with the client for
 // schema validation on calls to AddConstraint. It also returns a copy of the CRD describing
 // the constraint.
 func (c *Client) AddTemplate(ctx context.Context, templ *templates.ConstraintTemplate) (*types.Responses, error) {
 	resp := types.NewResponses()
-	crd, err := c.CreateCRD(ctx, templ)
-	if err != nil {
-		return resp, err
-	}
-
-	var src string
-	var target TargetHandler
-	for _, v := range templ.Spec.Targets {
-		k := v.Target
-		t, ok := c.targets[k]
-		if !ok {
-			return resp, fmt.Errorf("Target %s not recognized", k)
-		}
-		target = t
-		src = v.Rego
-	}
 
-	path := createTemplatePath(target.GetName(), crd.Spec.Names.Kind)
-	conformingSrc, err := c.rConformer.ensureRegoConformance(crd.Spec.Names.Kind, path, src)
+	artifacts, err := c.createTemplateArtifacts(templ)
 	if err != nil {
 		return resp, err
 	}
 
 	c.constraintsMux.Lock()
 	defer c.constraintsMux.Unlock()
-	if err := c.backend.driver.PutModule(ctx, path, conformingSrc); err != nil {
-		return resp, err
+
+	var rollbackPaths []string
+	defer func() {
+		for _, path := range rollbackPaths {
+			if _, err := c.backend.driver.DeleteModule(ctx, path); err != nil {
+				glog.Warningf("encountered error while rolling back module %s", path)
+			}
+		}
+	}()
+
+	for _, mod := range artifacts.modules {
+		if err := c.backend.driver.PutModule(ctx, mod.path, mod.rego); err != nil {
+			return resp, err
```

Comment:
Another problem with this AddModule loop is that currently the Gatekeeper update handler for constraint templates just calls `AddTemplate()` to replace the original code. That behavior is incompatible with this update as-written because it could lead to dangling library entries in the case where an updated template is removing a library (there is no tombstone for deletion).

One possible suggestion: if we have `PutModules()` (suggested in above comment) take a "root" argument, where everything under that root is replaced atomically, we can keep the old behavior. I'm open to other ideas, but that's what occurs to me at the moment.

One performance-enhancing item we can add to the local driver is to check for changes to the template source (if an overwrite) and only re-run the compiler if there is a delta (optional enhancement to add to this PR if we go with the above change).

Reply:
So it sounds like I need to add PutModules/DeleteModules, correct?  How do the following signatures look?
```
PutModules(ctx context.Context, namePrefix string, sources []string) error
DeleteModules(ctx context.Context, namePrefix string) (count int, err error)
```

Using a single key for many modules seems like it would be the right interface, I'll look into that.

As for the performance aspect, I think I'd like focus this change on correctness since we could always optimize later on.

Reply:
Re: performance SGTM, though in the local one, we should aim to run compiles at most once per call to `PutModules()`, as the inefficiency of doing otherwise is quite high and avoiding the multiple calls should be easy. Remote is out-of-scope, as that would require changes to the OPA REST interface.

My only concern with the prefix design is we are assuming that the consumers of PutModules() and PutModule() are aware of this clobbering-by-prefix potential. I'm fine with the risk as I think the CF should completely encapsulate these methods, but if there is an easy alternative it's worth pursuing.

Reply:
Added PutModules / DeleteModules function in local.go, PTAL and let me know if that impl is good then I can add some unit tests and integrate here.

Reply:
I should have closed this out earlier.